### ISSUE TITLE
problem with ncm-nfs/perl-CAF 15.2.0

### ISSUE THREAD

**Issue by nowack73** [quattor/configuration-modules-core#468]
Hi,

I am trying to upgrade from version 14.10.0 to 15.2.0. The partitioning did no change but the new component "fstab" produces the following error.

```text
[INFO] running component: fstab
---------------------------------------------------------
[VERB] Opening file /etc/fstab
[VERB] Updating /etc/fstab with new or modified contents
[VERB] Changes to /etc/fstab:

[VERB] File /etc/fstab was not modified
[ERROR] component fstab executing method Configure fails: Can't use an undefined value as a SCALAR reference at /usr/lib/perl/CAF/FileWriter.pm line 259, <GEN76> line 16.

[ERROR] cannot execute configure on component fstab

=========================================================
```

This is the configuration:
```text
# ncm-query --comp fstab

[INFO] Subtree: /software/components/fstab
+-fstab
  $ active : (boolean) 'true'
  +-dependencies
    +-pre
      $ 0 : (string) 'spma'
  $ dispatch : (boolean) 'true'
  +-protected_mounts
    $ 0 : (string) '/'
    $ 1 : (string) '/usr'
    $ 2 : (string) '/home'
    $ 3 : (string) '/opt'
    $ 4 : (string) '/var'
    $ 5 : (string) '/var/lib'
    $ 6 : (string) '/var/lib/rpm'
    $ 7 : (string) '/usr/bin'
    $ 8 : (string) '/usr/sbin'
    $ 9 : (string) '/usr/lib'
    $ 10 : (string) '/usr/local/bin'
    $ 11 : (string) '/usr/lib64'
    $ 12 : (string) '/bin'
    $ 13 : (string) '/sbin'
    $ 14 : (string) '/etc'
    $ 15 : (string) 'swap'
    $ 16 : (string) '/proc'
    $ 17 : (string) '/sys'
    $ 18 : (string) '/dev/pts'
    $ 19 : (string) '/dev/shm'
    $ 20 : (string) '/media/floppy'
    $ 21 : (string) '/mnt/floppy'
    $ 22 : (string) '/media/cdrecorder'
    $ 23 : (string) '/media/cdrom'
    $ 24 : (string) '/mnt/cdrom'
    $ 25 : (string) '/boot'
  +-register_change
    $ 0 : (string) '/system/filesystems'


# ncm-query --dump /system/filesystems

[INFO] Subtree: /system/filesystems
+-filesystems
  +-0
    $ block_device : (string) 'partitions/sda1'
    $ format : (boolean) 'true'
    $ freq : (long) '1'
    $ mkfsopts : (string) ''
    $ mount : (boolean) 'true'
    $ mountopts : (string) 'defaults'
    $ mountpoint : (string) '/boot'
    $ pass : (long) '2'
    $ preserve : (boolean) 'true'
    $ type : (string) 'ext3'
  +-1
    $ block_device : (string) 'partitions/sda2'
    $ format : (boolean) 'true'
    $ freq : (long) '1'
    $ mkfsopts : (string) ''
    $ mount : (boolean) 'true'
    $ mountopts : (string) 'defaults'
    $ mountpoint : (string) '/'
    $ pass : (long) '1'
    $ preserve : (boolean) 'true'
    $ type : (string) 'ext4'
  +-2
    $ block_device : (string) 'partitions/sda3'
    $ format : (boolean) 'true'
    $ freq : (long) '0'
    $ mkfsopts : (string) ''
    $ mount : (boolean) 'true'
    $ mountopts : (string) 'defaults'
    $ mountpoint : (string) 'swap'
    $ pass : (long) '0'
    $ preserve : (boolean) 'true'
    $ type : (string) 'swap'
  +-3
    $ block_device : (string) 'partitions/sda4'
    $ format : (boolean) 'false'
    $ freq : (long) '1'
    $ mkfsopts : (string) '-m 0'
    $ mount : (boolean) 'true'
    $ mountopts : (string) 'defaults'
    $ mountpoint : (string) '/user'
    $ pass : (long) '2'
    $ preserve : (boolean) 'true'
    $ type : (string) 'ext4'
```

Best regards,
  Andreas

**Comment by stdweird** [comment]
@nowack73 can you paste the lines around /usr/lib/perl/CAF/FileWriter.pm line 259? (this is the stringify in my CAF, but that makes no sense at all)
also, can you post the version of CAF and ncm-lib-blockdevices?
and is this very reproducable (and are you willing to run this through some debugger? needs some additional perl rpms, and ofcourse, not a very critical production system :smile: )

**Comment by nowack73** [comment]
Hi Stijn,

this is from /usr/lib/perl/CAF/FileWriter.pm:
```text
255: sub stringify
256:{
257:     my $self = shift;
258:     my $str = $self->string_ref;
259:     return $$str;
260:}
```

Both packages are up-to-date:
```
# rpm -q perl-CAF ncm-lib-blockdevices
perl-CAF-15.2.0-1.noarch
ncm-lib-blockdevices-15.2.0-1.noarch
```

It is reproducable and I can try to debug it.

Best regards,
  Andreas

**Comment by stdweird** [comment]
@nowack73 lets start with
```
yum install -y perl-App-cpanminus.noarch  perl-devel gcc perl-Test-Simple
cpanm install Devel::Trace
perl -t -d:Trace `which ncm-ncd` --co fstab >& fstab_output
```
and mail me the file :smile:

**Comment by stdweird** [comment]
@nowack73 thanks for the output
this looks like https://github.com/quattor/ncm-lib-blockdevices/issues/43
can you try with the rc1 rpm http://yum.quattor.org/testing/15.4.0-rc1/ncm-lib-blockdevices-15.4.0-rc1_1.noarch.rpm (or apply the commit mentioned in the ticket)

**Comment by stdweird** [comment]
@nowack73 can you confirm this is resolved in 15.4?

**Comment by stdweird** [comment]
tried to conatct @nowack73, but no reply eyt. closing this one since i'm quite sure it's fixed in the above commit. please reopen the issue if the problem remains
